File: send_command.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  bez nazwy.py
#  
#  Copyright 2013 Unknown <kaszak@localhost>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#  
import sys
import os
from time import sleep

FIFO = '/tmp/' + os.getenv('USER') + '/py3status.fifo'

class FileLock():
    def __init__(self, file_path, delay=.05):
        self._lock_path = file_path + '.lock'
        self._locked = False
        self.delay = delay
        
    def acquire(self):
        while True:
            try:
                self.descriptor = os.open(self._lock_path, flags=os.O_CREAT|os.O_EXCL|os.O_RDWR)
                break
            except OSError as fuck:
                if fuck.errno == 17: #  Means that the file already exist
                    sleep(self.delay)
                else:
                    raise
        self._locked = True
        
    def release(self):
        if self._locked:
            os.close(self.descriptor)
            os.unlink(self._lock_path)
            self._locked = False
            

def main():
    '''
    Simple tool to send commands to py3status.
    Usage:
    send_command.py TARGET COMMAND
    '''
    if len(sys.argv) < 3:
        sys.exit(1)
    
    target = sys.argv[1]
    command = ' '.join(sys.argv[2:])
    lock = FileLock(FIFO)
    lock.acquire()
    with open(FIFO, 'w') as fifo:
        fifo.write(target + ':' + command)
        fifo.flush()
    lock.release()

if __name__ == '__main__':
	main()

File: send_command.c
/*
 * send_command2.c
 * 
 * Copyright 2013 Kaszak <kaszak696@gmail.com>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <limits.h>
#include "lock.h"
#include "config.h"

/* Tool to pass comands trough FIFO pipe. */
    
int main(int argc, char **argv)
{   
    int len;                   
    char* filename;
    char* lockname;
    char* username;
    FILE* fp;
    Lock* lock;
    
    filename = (char*)calloc((len = strlen(TMPDIR) + strlen(username = getenv("USER")) + strlen(FILENAME) + 1), sizeof(char));
    lockname = (char*)calloc(len + strlen(LOCK_SUFFIX), sizeof(char));
    
    // Construct path
    strcat(filename, TMPDIR);
    strcat(filename, username);
    strcat(filename, FILENAME);
    
    // Ditto, for .lock file
    strcat(lockname, filename);
    strcat(lockname, LOCK_SUFFIX);
    
    // send_command TARGET COMMAND
    if(argc < 3) exit(1);
    
    // lock the file
    if((lock = acquire(lockname)) == NULL) exit(1);

    if((fp = fopen(filename, "w")) == NULL) 
    {
        release(lock);
        exit(1);
    }

    fprintf(fp, "%s:%s", argv[1], argv[2]);

    fclose(fp);
    free(filename);
    free(lockname);
    
    // remove the .lock file and bail out
    release(lock);
    return 0;
}
